### PR TITLE
[Fix #1913] Allow toggling of current buffer connection

### LINKED ISSUE #1913: cider-toggle-buffer-connection fails
## Expected behavior

`cider-toggle-buffer-connection` should toggle between CLJ and CLJS connections (if both exist) in a `clojurec-mode` buffer (e.g. `cljc` files) or in a `cider-clojure-interaction-mode` buffer (e.g. `*cider-scratch*`).  This should allow expressions to be evaluated in the appropriate connection. Also, `cider-connections` should still contain both CLJ and CLJS connections after toggling.

## Actual behavior

Connection to REPL is seemingly lost; evaluating expressions causes CIDER to complain that a Clojure REPL is required and that a jack in is probably needed.  Also, `cider-connections` loses one of the connections (the current one).

## Steps to reproduce the problem

1. Do `lein new cider-toggle-buffer-connection`
2. Edit `project.clj` to the following:
``` clojure
(defproject cider-toggle-buffer-connection "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [com.cemerick/piggieback "0.2.1"]
                 [org.clojure/clojure "1.7.0"]]
  :repl-options {:nrepl-middleware [cemerick.piggieback/wrap-cljs-repl]}
  )
```
3. Do `mv src/cider-toggle-buffer-connection/core.cljs src/cider-toggle-buffer-connection/core.cljc`
4. Edit `src/cider-toggle-buffer-connection/core.cljc` to the following:
``` clojure
(ns cider-toggle-buffer-connection.core)
(println "Test")
```
5. Open `src/cider-toggle-buffer-connection/core.cljc` in emacs.
6. Do `cider-jack-in-clojurescript`.
7. Take note of the value of `cider-connections`: there should be two connections (CLJ and CLJS).
8. Evaluate `(println "Test")` with `cider-eval-last-sexp`.  This should print "Test" in the `*cider-repl cider-toggle-buffer-connection*` buffer.
9. Do `cider-toggle-buffer-connection`.
10. Evaluate `(println "Test")` with `cider-eval-last-sexp`.  This should cause CIDER to complain with the following message:
> user-error: ‘cider-eval-last-sexp’ needs a Clojure REPL.
If you don’t know what that means, you probably need to jack-in (‘C-c M-j’).
11. Take note of the value of `cider-connections`: only the CLJS connections remains.

### CIDER version information

```
;; CIDER 0.14.0 (Berlin), nREPL 0.2.12
;; Clojure 1.8.0, Java 1.8.0_74
```

### Emacs version

> GNU Emacs 25.1.3 (x86_64-apple-darwin15.6.0, Carbon Version 157 AppKit 1404.47) of 2016-09-18

### Operating system

OSX 10.11.6

### PR DESCRIPTION
**Pull request by dpsutton**
Cljc buffers send their evals to both clj and cljs repls if available due to
`cider-map-connections`. Toggling a current buffer's connection involves hiding
the other connection.

Previously, when toggling _again_, the original list was not consulted and only
the truncated list, preventing the other connection from being found. This
allows for the full list to be searched for the other buffer.

In addition, a prefix dictates that the local connection list is discarded in
favor of the full list, restoring the evaluation in both clj and cljs
buffers (if both are present).

- [x] The commits are consistent with our [contribution guidelines][1]
- [x] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

See https://github.com/clojure-emacs/cider/issues/1913

### CODE REVIEW

**Comment by dpsutton** [comment]
> The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends --force-yes install autogen ca-certificates curl gcc libgmp-dev m4 make pkg-config xz-utils" failed and exited with 100 during .

> An error occurred while generating the build script.

Seems like two transient build errors in travis

File: cider-client.el [hunk 1]
```diff
@@ -225,14 +225,35 @@ such a link cannot be established automatically."
...
+  (interactive "P")
   (cider-ensure-connected)
-  (let ((other-conn (cider-other-connection)))
-    (if other-conn
-        (setq-local cider-connections (list other-conn))
-      (user-error "No other connection available"))))
+  (if restore-all
+      (progn
+        (kill-local-variable 'cider-connections)
+        (let ((types (mapcar #'cider--connection-type (cider-connections))))
```
Comment (bbatsov) [inline cider-client.el]: I don't get what's the purpose of this portion of the code.

File: cider-client.el [hunk 2]
```diff
@@ -225,14 +225,35 @@ such a link cannot be established automatically."
...
+          (message (format "Cider connections available: %s" types))))
+    (let ((current (cider-current-connection))
```
Comment (bbatsov) [inline cider-client.el]: I'd name this `current-conn`.

**Comment by bbatsov** [comment]
> Cljc buffers send their evals to both clj and cljs repls if available due to
cider-map-connections. Toggling a current buffer's connection involves hiding
the other connection.

This reminds me it might be nice to explain this behavior somewhere in the manual.

File: cider-client.el (hunk 1 above)
Comment (dpsutton) [reply cider-client.el]: The default behavior of a `cljc` buffer is to map evaluations across all connections. Toggling switches to just one or the other. With a prefix arg, (`restore-all`), it deletes the buffer local list of connections and allows the eval mechanism to see the original list of connections.

File: cider-client.el (hunk 2 above)
Comment (dpsutton) [reply cider-client.el]: done and push.

**Comment by dpsutton** [comment]
documentation updated in `up_and_running.md`

File: cider-client.el (hunk 1 above)
Comment (bbatsov) [reply cider-client.el]: I get what the code does in general - I was referring to this `let` and the message inside it. You get the types of all connections and you print them in a message basically, which seems strange to me.

Comment (dpsutton) [reply cider-client.el]: Ah yes. The connections stuff can be rather opaque. So when toggling it it tells you which connection it is going to use, and then I give a message listing all connections when all of the connections are "untoggled"

Comment (bbatsov) [reply cider-client.el]: Ah, I finally get it.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**